File: ios/Halogen/Features/Podcasts/PodcastEditView.swift
import SwiftUI

/// Edit a podcast's own fields (title / description / feed URL) — the web's
/// `/podcasts/:id/edit`.
struct PodcastEditView: View {
    let core: HalogenCore
    let podcast: PodcastData

    @Environment(\.dismiss) private var dismiss
    @State private var title: String
    @State private var description: String
    @State private var feedUrl: String
    @State private var error: String?
    @State private var saving = false

    init(core: HalogenCore, podcast: PodcastData) {
        self.core = core
        self.podcast = podcast
        _title = State(initialValue: podcast.title)
        _description = State(initialValue: podcast.description)
        _feedUrl = State(initialValue: podcast.feed_url)
    }

    var body: some View {
        Form {
            TextField("Title", text: $title)
            TextField("Feed URL", text: $feedUrl)
                .textContentType(.URL)
                .keyboardType(.URL)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
            Section("Description") {
                TextEditor(text: $description).frame(minHeight: 100)
            }
            if let error {
                Text(error).font(.footnote).foregroundStyle(.red)
            }
            Button {
                Task { await save() }
            } label: {
                if saving {
                    ProgressView().frame(maxWidth: .infinity)
                } else {
                    // Online-only (feed-URL edits must surface server
                    // validation) — say so instead of failing raw (web
                    // podcast_edit.rs: submit disabled + labelled offline).
                    Text(core.isOffline ? "Offline — reconnect to save" : "Save")
                        .frame(maxWidth: .infinity)
                }
            }
            .disabled(
                saving || core.isOffline
                    || title.trimmingCharacters(in: .whitespaces).isEmpty)
        }
        .navigationTitle("Edit podcast")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func save() async {
        saving = true
        defer { saving = false }
        do {
            try await core.updatePodcast(
                id: podcast.id,
                title: title,
                description: description.isEmpty ? nil : description,
                feedUrl: feedUrl
            )
            await core.models?.podcasts.refresh()
            dismiss()
        } catch {
            self.error = FriendlyError.message(error)
        }
    }
}

File: android/app/src/main/kotlin/org/fgsec/halogen/features/podcasts/PodcastEditView.kt
package org.fgsec.halogen.features.podcasts

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.halogenIcon
import org.fgsec.halogen.core.FriendlyError
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.wire.PodcastData

/// Edit a podcast's own fields (title / description / feed URL) — the web's
/// `/podcasts/:id/edit`.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PodcastEditView(core: HalogenCore, podcast: PodcastData, onBack: (() -> Unit)? = null) {
    var title by remember { mutableStateOf(podcast.title) }
    var description by remember { mutableStateOf(podcast.description) }
    var feedUrl by remember { mutableStateOf(podcast.feed_url) }
    var error by remember { mutableStateOf<String?>(null) }
    var saving by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    suspend fun save() {
        saving = true
        try {
            core.updatePodcast(
                id = podcast.id,
                title = title,
                description = description.ifEmpty { null },
                feedUrl = feedUrl,
            )
            core.models?.podcasts?.refresh()
            onBack?.invoke()
        } catch (e: Exception) {
            error = FriendlyError.message(e)
        } finally {
            saving = false
        }
    }

    Scaffold(
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Edit podcast") },
                navigationIcon = {
                    if (onBack != null) {
                        IconButton(onClick = onBack) {
                            Icon(halogenIcon("chevron.left"), contentDescription = "Back")
                        }
                    }
                },
            )
        },
    ) { padding ->
        Column(
            Modifier.fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            OutlinedTextField(
                value = title,
                onValueChange = { title = it },
                label = { Text("Title") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
            )
            OutlinedTextField(
                value = feedUrl,
                onValueChange = { feedUrl = it },
                label = { Text("Feed URL") },
                singleLine = true,
                keyboardOptions = KeyboardOptions(
                    capitalization = KeyboardCapitalization.None,
                    autoCorrectEnabled = false,
                    keyboardType = KeyboardType.Uri,
                ),
                modifier = Modifier.fillMaxWidth(),
            )
            OutlinedTextField(
                value = description,
                onValueChange = { description = it },
                label = { Text("Description") },
                modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 100.dp),
            )
            error?.let {
                Text(
                    it,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.error,
                )
            }
            Button(
                onClick = { scope.launch { save() } },
                enabled = !saving && !core.isOffline && title.trim().isNotEmpty(),
                modifier = Modifier.fillMaxWidth(),
            ) {
                if (saving) {
                    CircularProgressIndicator(modifier = Modifier.size(18.dp), strokeWidth = 2.dp)
                } else {
                    // Online-only (feed-URL edits must surface server
                    // validation) — say so instead of failing raw (web
                    // podcast_edit.rs: submit disabled + labelled offline).
                    Text(if (core.isOffline) "Offline — reconnect to save" else "Save")
                }
            }
        }
    }
}
